### PR TITLE
Don't check the name of a destructor during parsing.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This PR specifically makes it so that the name of a destructor is not checked (while parsing). Previously, we disallowed (at parse time) the destructor name from not matching the type name. There's no reason for this complexity and specificity to be at the parser level. Syntactically we know what's going on, and we can allow this while parsing and check and report the error later.  It also prevents incrementally reusing a destructor when a type name changes.

Tagging @dotnet/roslyn-compiler @gafter @VSadov

This part of a long term effort to remove unnecessary errors from the parser and move them to later stages of the compiler pipeline. Doing this makes the parser simpler, makes it faster to get a syntax tree, and makes our incremental implementation simpler (because the parser/incremental-parser needs to track less contextual states).

This is also important as syntax-diagnostics end up causing issues for higher level IDE features.

**Comment by TyOverby** [comment]
👍

**Comment by AlekseyTs** [comment]
It is very hard to distinguish these PRs between each other, because they all have title starting with: "Move diagnostic from parser to later phase." I would like to suggest to actually state what the PR is about in the title, rather than using a common theme. For example, this PR could have title: "Do not check destructor's name during parsing."

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -2765,20 +2765,12 @@ private DestructorDeclarationSyntax ParseDestructorDeclaration(string typeName,
...
             var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
 
-            BlockSyntax body;
-            ArrowExpressionClauseSyntax expressionBody;
-            SyntaxToken semicolon;
-            this.ParseBlockAndExpressionBodiesWithSemicolon(out body, out expressionBody, out semicolon,
-                                                            requestedExpressionBodyFeature: MessageID.IDS_FeatureExpressionBodiedDeOrConstructor);
-
+            this.ParseBlockAndExpressionBodiesWithSemicolon(
+                out var body, out var expressionBody, out var semicolon,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >var [](start = 20, length = 3)

Please replace "var"s with types or revert the change to this statement.
